File: src/FileProcessor.ts
import fs from "fs/promises";
import path from "path";
import Database from "better-sqlite3";
import { BrowserWindow } from "electron";
import { statSync } from "fs";
import fetch from "node-fetch";
import { getCategoryFromExtension } from "./lib/utils";
import { FileMetadata, SearchSectionType } from "./types";
import mammoth from "mammoth";
import pdfParse from "pdf-parse";
import textract from "textract";

export default class FileProcessor {
  private db: Database.Database;
  private mainWindow: BrowserWindow | null;
  private totalFiles: number = 0;
  private processedFiles: number = 0;

  constructor(db: Database.Database, mainWindow: BrowserWindow | null = null) {
    this.db = db;
    this.mainWindow = mainWindow;
  }

  /* processes the path provided by the user. The path can be a directory or a single file. If a directory, it will get all of the files in that directory and then index it and create an embedding from it and save it to the database. */
  public async processPaths(
    paths: string[]
  ): Promise<{ success: boolean; totalFiles: number }> {
    try {
      this.totalFiles = 0;
      this.processedFiles = 0;

      // collect all files from the provided paths.
      let allFiles: FileMetadata[] = [];

      for (const targetPath of paths) {
        if (this.isDirectory(targetPath)) {
          const files = await this.getAllFiles(targetPath);
          allFiles = allFiles.concat(files);
        } else {
          // If it's a file, get its stats.
          const stats = await fs.stat(targetPath);
          allFiles.push({
            path: targetPath,
            name: path.basename(targetPath),
            extension: path.extname(targetPath),
            size: stats.size,
            type: SearchSectionType.Files,
          });
        }
      }

      this.totalFiles = allFiles.length;
      this.updateProgress();

      // Process each file by generating embedding and indexing metadata.
      for (const file of allFiles) {
        try {
          await this.processFile(file);
          this.processedFiles++;
          this.updateProgress();
        } catch (e) {
          console.error("Failed to process file:", file.path, e);
        }
      }
      return { success: true, totalFiles: this.processedFiles };
    } catch (error) {
      console.error("Error processing paths:", error);
      throw error;
    }
  }

  /**
   * Processes a single file: extracts text, generates an embedding, and updates SQLite.
   */
  private async processFile(file: FileMetadata): Promise<void> {
    const ext = path.extname(file.path);
    const category = getCategoryFromExtension(ext);
    const name = path.basename(file.path);
    const content = await this.extractText(file.path);

    if (!content) {
      return;
    }

    try {
      // First check if file exists in database
      const checkStmt = this.db.prepare("SELECT id FROM files WHERE path = ?");
      const existingFile = checkStmt.get(file.path) as
        | { id: number }
        | undefined;

      if (existingFile) {
        // File exists, update file metadata
        const updateFileStmt = this.db.prepare(`
          UPDATE files 
          SET name = ?, 
              category = ?, 
              updated_at = CURRENT_TIMESTAMP 
          WHERE id = ?
        `);

        updateFileStmt.run(name, category, existingFile.id);

        // Generate new embedding and update embeddings table
        const embedding = await this.addFileEmbedding(existingFile.id, content);
        const embeddingJson = JSON.stringify(embedding);

        const updateEmbeddingStmt = this.db.prepare(`
          INSERT OR REPLACE INTO embeddings (
            file_id, 
            embedding, 
            updated_at
          ) VALUES (?, ?, CURRENT_TIMESTAMP)
        `);

        updateEmbeddingStmt.run(existingFile.id, embeddingJson);
      } else {
        // File doesn't exist, insert file first to get ID
        const insertFileStmt = this.db.prepare(`
          INSERT INTO files (
            path, 
            name, 
            category, 
            extension,
            created_at,
            updated_at
          ) VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
        `);

        const result = insertFileStmt.run(
          file.path,
          name,
          category,
          file.extension
        );
        const newFileId = result.lastInsertRowid as number;

        // Now generate embedding and insert into embeddings table
        const embedding = await this.addFileEmbedding(newFileId, content);
        const embeddingJson = JSON.stringify(embedding);

        const insertEmbeddingStmt = this.db.prepare(`
          INSERT INTO embeddings (
            file_id, 
            embedding,
            created_at,
            updated_at
          ) VALUES (?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
        `);

        insertEmbeddingStmt.run(newFileId, embeddingJson);
      }
    } catch (error) {
      console.error(`Error processing file ${file.path}:`, error);
      throw error;
    }
  }

  /**
   * Indexes a single file's metadata in SQLite.
   * This function does NOT generate an embedding; it only updates the file metadata.
   */
  public async indexFile(filePath: string): Promise<void> {
    const ext = path.extname(filePath);
    const category = getCategoryFromExtension(ext);
    const name = path.basename(filePath);
    const stats = await fs.stat(filePath);
    const updateStmt = this.db.prepare(`
        UPDATE files 
        SET name = ?, category = ?, updated_at = CURRENT_TIMESTAMP 
        WHERE path = ?
      `);
    const result = updateStmt.run(name, category, filePath);
    if (result.changes === 0) {
      const insertStmt = this.db.prepare(`
          INSERT INTO files (path, name, category) VALUES (?, ?, ?)
        `);
      insertStmt.run(filePath, name, category);
    }
  }

  /* 
  Creates an embedding for a single file and updates the database with that embedding
  */
  public async createEmbeddingForFile(filePath: string): Promise<number[]> {
    const content = await this.extractText(filePath);
    if (content) {
      return await this.getEmbedding(content);
    }
    return [];
  }

  // gets all of the files for a given directory path by recursively
  // searching through the directory
  private async getAllFiles(dirPath: string): Promise<FileMetadata[]> {
    const files: FileMetadata[] = [];

    try {
      const entries = await fs.readdir(dirPath, {
        withFileTypes: true,
      });

      for (const entry of entries) {
        const fullPath = path.join(dirPath, entry.name);

        if (entry.isDirectory()) {
          // Recursively get files from subdirectories
          const subFiles = await this.getAllFiles(fullPath);
          files.push(...subFiles);
        } else {
          const stats = await fs.stat(fullPath);
          files.push({
            path: fullPath,
            name: entry.name,
            extension: path.extname(entry.name),
            size: stats.size,
            type: SearchSectionType.Files,
          });
        }
      }
    } catch (error) {
      console.error(`Error reading directory ${dirPath}:`, error);
    }

    return files;
  }

  /**
   * Extracts text from a file. Currently supports only .txt files.
   */
  private async extractText(filePath: string): Promise<string> {
    const ext = path.extname(filePath).toLowerCase();

    // Extensions we can treat as plain text.
    const plainTextExtensions = new Set([
      ".txt",
      ".js",
      ".ts",
      ".jsx",
      ".tsx",
      ".py",
      ".java",
      ".cpp",
      ".html",
      ".css",
      ".json",
      ".xml",
      ".yaml",
      ".yml",
    ]);

    if (plainTextExtensions.has(ext)) {
      return this.extractTextFromPlain(filePath);
    } else if (ext === ".docx") {
      return this.extractTextFromDocx(filePath);
    } else if (ext === ".pdf") {
      return this.extractTextFromPDF(filePath);
    } else if (ext === ".doc" || ext === ".rtf") {
      return this.extractTextFromDocOrRtf(filePath);
    }

    // If extension is not supported, return an empty string.
    return "";
  }

  async extractTextFromDocOrRtf(filePath: string): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      textract.fromFileWithPath(filePath, (err, text) => {
        if (err) {
          console.error(`Error extracting DOC/RTF file ${filePath}:`, err);
          reject(err);
        } else {
          resolve(text);
        }
      });
    }).catch(() => "");
  }

  async extractTextFromPDF(filePath: string): Promise<string> {
    try {
      const dataBuffer = await fs.readFile(filePath);
      const data = await pdfParse(dataBuffer);
      return data.text;
    } catch (error) {
      console.error(`Error extracting PDF file ${filePath}:`, error);
      return "";
    }
  }

  async extractTextFromDocx(filePath: string): Promise<string> {
    try {
      const result = await mammoth.extractRawText({ path: filePath });
      return result.value;
    } catch (error) {
      console.error(`Error extracting DOCX file ${filePath}:`, error);
      return "";
    }
  }

  async extractTextFromPlain(filePath: string): Promise<string> {
    try {
      return await fs.readFile(filePath, "utf-8");
    } catch (error) {
      console.error(`Error reading plain text file ${filePath}:`, error);
      return "";
    }
  }

  /**
   * Calls the Python microservice to generate an embedding from text.
   */
  private async getEmbedding(text: string): Promise<number[]> {
    const response = await fetch("http://127.0.0.1:8000/embed", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text }),
    });
    if (!response.ok) {
      throw new Error("Failed to get embedding");
    }
    const data = await response.json();
    return data.embedding;
  }

  /**
   * Generates an embedding for the given text and then adds or updates that file's embedding
   * in the vector index via the /add_file endpoint.
   *
   * @param fileId - A unique identifier for the file (managed by your app)
   * @param text - The text content extracted from the file
   * @returns The embedding vector as an array of numbers
   */
  private async addFileEmbedding(
    fileId: number,
    text: string
  ): Promise<number[]> {
    const embedding = await this.getEmbedding(text);

    // update the vector index
    const response = await fetch("http://127.0.0.1:8000/add_file", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ file_id: fileId ?? 0, embedding }),
    });
    if (!response.ok) {
      throw new Error("Failed to add file embedding");
    }
    return embedding;
  }

  private updateProgress(): void {
    if (this.mainWindow) {
      this.mainWindow.webContents.send("indexing-progress", {
        total: this.totalFiles,
        processed: this.processedFiles,
        percentage: Math.round((this.processedFiles / this.totalFiles) * 100),
      });
    }
  }

  // returns a boolean if the provided path is a directory
  public isDirectory(path: string): boolean {
    if (!path || typeof path !== "string") {
      return false;
    }

    try {
      return statSync(path).isDirectory();
    } catch (error) {
      if (
        error instanceof Error &&
        "code" in error &&
        error.code === "ENOENT"
      ) {
        return false;
      }
      throw error;
    }
  }
}

File: go-orchestrator/file-processor/file-processor.go
package fileprocessor

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/ledongthuc/pdf"
	_ "github.com/mattn/go-sqlite3"
	"github.com/unidoc/unioffice/document"
)

type SearchSectionType string

type SearchCategory string

const (
	Files    SearchSectionType = "files"
	Apps     SearchSectionType = "apps"
	Semantic SearchSectionType = "semantic"
)

const (
	CategoryApplications SearchCategory = "Applications"
	CategoryPDFDocuments SearchCategory = "PDF Documents"
	CategoryDocuments    SearchCategory = "Documents"
	CategoryImages       SearchCategory = "Images"
	CategorySpreadsheets SearchCategory = "Spreadsheets"
	CategoryOther        SearchCategory = "Other"
)

type BaseMetadata struct {
	ID   *int   `json:"id,omitempty"`
	Name string `json:"name"`
	Path string `json:"path"`
}

type FileMetadata struct {
	BaseMetadata
	Type      SearchSectionType `json:"type"`
	Extension string            `json:"extension"`
	Size      int64             `json:"size"`
	UpdatedAt *string           `json:"updated_at,omitempty"`
	CreatedAt *string           `json:"created_at,omitempty"`
}

type AppMetadata struct {
	BaseMetadata
	Type        SearchSectionType `json:"type"`
	IsRunning   bool              `json:"isRunning"`
	MemoryUsage *float64          `json:"memoryUsage,omitempty"`
	CPUUsage    *float64          `json:"cpuUsage,omitempty"`
	IconDataURL *string           `json:"iconDataUrl,omitempty"`
}

type SemanticMetadata struct {
	BaseMetadata
	Type      SearchSectionType `json:"type"`
	Extension string            `json:"extension"`
	Distance  float64           `json:"distance"`
	Content   *string           `json:"content,omitempty"`
}

// handles all file processing operations
type FileProcessor struct {
	Db             *sql.DB
	TotalFiles     int
	ProcessedFiles int
	mu             sync.Mutex
	wg             sync.WaitGroup
	semaphore      chan struct{} // for limiting concurrent operations
}

type ProcessingStatus struct {
	Total      int `json:"total"`
	Processed  int `json:"processed"`
	Percentage int `json:"percentage"`
}

// initializes a new file processor
func NewFileProcessor(dbPath string) (*FileProcessor, error) {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %v", err)
	}

	return &FileProcessor{
		Db:        db,
		semaphore: make(chan struct{}, 4), // limit to 4 concurrent ops, but we can be smarter about this
	}, nil
}

// ProcessPaths processes multiple file paths concurrently
func (fp *FileProcessor) ProcessPaths(paths []string) (map[string]interface{}, error) {
	fp.TotalFiles = 0
	fp.ProcessedFiles = 0
	var allFiles []FileMetadata

	// Collect all files first
	for _, targetPath := range paths {
		if fp.isDirectory(targetPath) {
			files, err := fp.getAllFiles(targetPath)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error processing directory %s: %v\n", targetPath, err)
				continue
			}
			allFiles = append(allFiles, files...)
		} else {
			info, err := os.Stat(targetPath)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error getting file info %s: %v\n", targetPath, err)
				continue
			}
			allFiles = append(allFiles, FileMetadata{
				BaseMetadata: BaseMetadata{
					Path: targetPath,
					Name: filepath.Base(targetPath),
				},
				Type:      Files,
				Extension: filepath.Ext(targetPath),
				Size:      info.Size(),
			})
		}
	}

	fp.TotalFiles = len(allFiles)
	fp.updateProgress()

	errChan := make(chan error, len(allFiles)) // creates a buffer channel to hold errors from all concurrent runs, len set to all files in case all files have an error

	for _, file := range allFiles {
		fp.wg.Add(1)
		go func(f FileMetadata) {
			defer fp.wg.Done()                // defers until the function is done executing
			fp.semaphore <- struct{}{}        //  used to manage concurrency, sends empty struct to semaphore, if full then it can't send it and will wait until it can
			defer func() { <-fp.semaphore }() // release semaphore making space for another go routine to start

			if err := fp.processFile(f); err != nil {
				errChan <- fmt.Errorf("error processing %s: %v", f.Path, err) // send error to errChan with path
				return
			}

			fp.mu.Lock()        // lock the struct
			fp.ProcessedFiles++ // increment processed files
			fp.updateProgress() // update the progress
			fp.mu.Unlock()      // unlock the struct
		}(file)
	}

	// Wait for all goroutines to complete
	fp.wg.Wait()
	close(errChan)

	// Collect any errors
	var errors []string
	for err := range errChan {
		errors = append(errors, err.Error())
	}

	result := map[string]interface{}{
		"success":    len(errors) == 0,
		"totalFiles": fp.ProcessedFiles,
	}
	if len(errors) > 0 {
		result["errors"] = errors
	}

	return result, nil
}

// processFile handles the processing of a single file
func (fp *FileProcessor) processFile(file FileMetadata) error {
	content, err := fp.extractText(file.Path)
	if err != nil {
		return fmt.Errorf("failed to extract text: %v", err)
	}
	if content == "" {
		return nil
	}

	category := getCategoryFromExtension(file.Extension)

	// start a transaction
	tx, err := fp.Db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	// Check if file exists
	var fileID int64
	err = tx.QueryRow("SELECT id FROM files WHERE path = ?", file.Path).Scan(&fileID)
	if err == sql.ErrNoRows {
		// Insert new file
		result, err := tx.Exec(`
			INSERT INTO files (path, name, category, extension, created_at, updated_at)
			VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
			file.Path, file.Name, category, file.Extension)
		if err != nil {
			return err
		}
		fileID, _ = result.LastInsertId()
	} else if err != nil {
		return err
	} else {
		// Update existing file
		_, err = tx.Exec(`
			UPDATE files 
			SET name = ?, category = ?, updated_at = CURRENT_TIMESTAMP 
			WHERE id = ?`,
			file.Name, category, fileID)
		if err != nil {
			return err
		}
	}

	// Generate embedding
	embedding, err := fp.getEmbedding(content)
	if err != nil {
		return err
	}

	embeddingJSON, err := json.Marshal(embedding)
	if err != nil {
		return err
	}

	// Update embedding
	_, err = tx.Exec(`
		INSERT OR REPLACE INTO embeddings (file_id, embedding, updated_at)
		VALUES (?, ?, CURRENT_TIMESTAMP)`,
		fileID, string(embeddingJSON))
	if err != nil {
		return err
	}

	// Update vector index
	err = fp.updateVectorIndex(fileID, embedding)
	if err != nil {
		return err
	}

	// commit the transaction
	return tx.Commit()
}

func (fp *FileProcessor) extractText(filePath string) (string, error) {
	ext := strings.ToLower(filepath.Ext(filePath))

	switch {
	case isPlainText(ext):
		return fp.extractTextFromPlain(filePath)
	case ext == ".pdf":
		return fp.extractTextFromPDF(filePath)
	case ext == ".docx":
		return fp.extractTextFromDOCX(filePath)
	default:
		return "", nil
	}
}

func isPlainText(ext string) bool {
	plainTextExts := map[string]bool{
		".txt": true, ".js": true, ".ts": true,
		".jsx": true, ".tsx": true, ".py": true,
		".java": true, ".cpp": true, ".html": true,
		".css": true, ".json": true, ".xml": true,
		".yaml": true, ".yml": true,
	}
	return plainTextExts[ext]
}

func (fp *FileProcessor) extractTextFromPlain(filePath string) (string, error) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func (fp *FileProcessor) extractTextFromPDF(filePath string) (string, error) {
	f, r, err := pdf.Open(filePath)
	if err != nil {
		return "", err
	}
	defer f.Close()

	var buf bytes.Buffer
	b, err := r.GetPlainText()
	if err != nil {
		return "", err
	}
	_, err = buf.ReadFrom(b)
	if err != nil {
		return "", err
	}
	return buf.String(), nil
}

func (fp *FileProcessor) extractTextFromDOCX(filePath string) (string, error) {
	doc, err := document.Open(filePath)
	if err != nil {
		return "", err
	}
	var text strings.Builder
	for _, para := range doc.Paragraphs() {
		for _, run := range para.Runs() {
			text.WriteString(run.Text())
		}
		text.WriteString("\n")
	}
	return text.String(), nil
}

// getEmbedding gets embedding from the Python microservice
func (fp *FileProcessor) getEmbedding(text string) ([]float64, error) {
	payload := map[string]string{"text": text}
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	resp, err := http.Post("http://127.0.0.1:8000/embed", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result struct {
		Embedding []float64 `json:"embedding"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result.Embedding, nil
}

// updateVectorIndex updates the vector index in the Python microservice
func (fp *FileProcessor) updateVectorIndex(fileID int64, embedding []float64) error {
	payload := map[string]interface{}{
		"file_id":   fileID,
		"embedding": embedding,
	}
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	resp, err := http.Post("http://127.0.0.1:8000/add_file", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to update vector index: %s", body)
	}

	return nil
}

// checks if the path is a directory
func (f *FileProcessor) isDirectory(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return info.IsDir()
}

// getAllFiles recursively gets all files in a directory
func (fp *FileProcessor) getAllFiles(dirPath string) ([]FileMetadata, error) {
	var files []FileMetadata
	err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() {
			files = append(files, FileMetadata{
				BaseMetadata: BaseMetadata{
					Path: path,
					Name: info.Name(),
				},
				Type:      Files,
				Extension: filepath.Ext(path),
				Size:      info.Size(),
			})
		}
		return nil
	})
	return files, err
}

// updateProgress prints the current progress to stdout for the Electron app to read
func (fp *FileProcessor) updateProgress() {
	if fp.TotalFiles > 0 {
		status := ProcessingStatus{
			Total:      fp.TotalFiles,
			Processed:  fp.ProcessedFiles,
			Percentage: int((float64(fp.ProcessedFiles) / float64(fp.TotalFiles)) * 100),
		}
		json.NewEncoder(os.Stdout).Encode(status)
	}
}

func getCategoryFromExtension(extension string) SearchCategory {
	if extension == "" {
		return CategoryOther
	}

	switch strings.ToLower(extension) {
	case ".app", ".exe", ".dmg":
		return CategoryApplications

	case ".pdf":
		return CategoryPDFDocuments

	case ".doc", ".docx", ".txt", ".rtf":
		return CategoryDocuments

	case ".jpg", ".jpeg", ".png", ".gif", ".svg", ".webp":
		return CategoryImages

	case ".js", ".ts", ".jsx", ".tsx", ".py", ".java", ".cpp",
		".html", ".css", ".json", ".xml", ".yaml", ".yml":
		return CategoryDocuments

	case ".mp4", ".mov", ".avi", ".mkv":
		return CategoryOther

	case ".mp3", ".wav", ".flac", ".m4a":
		return CategoryOther

	case ".xlsx", ".xls", ".csv":
		return CategorySpreadsheets

	case ".zip", ".rar", ".7z", ".tar", ".gz":
		return CategoryOther

	default:
		return CategoryOther
	}
}
